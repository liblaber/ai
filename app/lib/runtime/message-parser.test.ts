// Mock dependencies that might cause issues
jest.mock('~/utils/logger', () => ({
  createScopedLogger: () => ({
    trace: jest.fn(),
    debug: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
  }),
}));

jest.mock('~/utils/unreachable', () => ({
  unreachable: jest.fn(),
}));

// Import the actual class
import { StreamingMessageParser } from './message-parser';

interface ExpectedResult {
  output: string;
  callbacks?: {
    onArtifactOpen?: number;
    onArtifactClose?: number;
    onActionOpen?: number;
    onActionClose?: number;
  };
}

describe('StreamingMessageParser', () => {
  it('should pass through normal text', () => {
    const parser = new StreamingMessageParser();
    expect(parser.parse('test_id', 'Hello, world!')).toBe('Hello, world!');
  });

  it('should allow normal HTML tags', () => {
    const parser = new StreamingMessageParser();
    expect(parser.parse('test_id', 'Hello <strong>world</strong>!')).toBe('Hello <strong>world</strong>!');
  });

  describe('no artifacts', () => {
    it.each<[string | string[], ExpectedResult | string]>([
      ['Foo bar', 'Foo bar'],
      ['Foo bar <', 'Foo bar '],
      ['Foo bar <p', 'Foo bar <p'],
      [['Foo bar <', 's', 'p', 'an>some text</span>'], 'an>some text</span>'],
    ])('should correctly parse chunks and strip out liblab artifacts (%#)', (input, expected) => {
      runTest(input, expected);
    });
  });

  describe('valid artifacts without actions', () => {
    it.each<[string | string[], ExpectedResult | string]>([
      [
        'Some text before <liblabArtifact title="Some title" id="artifact_1">foo bar</liblabArtifact> Some more text',
        {
          output: 'Some text before  Some more text',
          callbacks: { onArtifactOpen: 1, onArtifactClose: 1, onActionOpen: 0, onActionClose: 0 },
        },
      ],
      [
        'Before <liblabArtifact title="Some title" id="artifact_1">foo</liblabArtifact> After',
        {
          output: 'Before  After',
          callbacks: { onArtifactOpen: 1, onArtifactClose: 1, onActionOpen: 0, onActionClose: 0 },
        },
      ],
    ])('should correctly parse chunks and strip out liblab artifacts (%#)', (input, expected) => {
      runTest(input, expected);
    });
  });

  describe('valid artifacts with actions', () => {
    it.each<[string | string[], ExpectedResult | string]>([
      [
        'Before <liblabArtifact title="Some title" id="artifact_1"><liblabAction type="shell">npm install</liblabAction></liblabArtifact> After',
        {
          output: 'Before  After',
          callbacks: { onArtifactOpen: 1, onArtifactClose: 1, onActionOpen: 1, onActionClose: 1 },
        },
      ],
    ])('should correctly parse chunks and strip out liblab artifacts (%#)', (input, expected) => {
      runTest(input, expected);
    });
  });
});

function runTest(input: string | string[], outputOrExpectedResult: string | ExpectedResult) {
  let expected: ExpectedResult;

  if (typeof outputOrExpectedResult === 'string') {
    expected = { output: outputOrExpectedResult };
  } else {
    expected = outputOrExpectedResult;
  }

  const callbacks = {
    onArtifactOpen: jest.fn((data: any) => {
      expect(data).toMatchSnapshot('onArtifactOpen');
    }),
    onArtifactClose: jest.fn((data: any) => {
      expect(data).toMatchSnapshot('onArtifactClose');
    }),
    onActionOpen: jest.fn((data: any) => {
      expect(data).toMatchSnapshot('onActionOpen');
    }),
    onActionClose: jest.fn((data: any) => {
      expect(data).toMatchSnapshot('onActionClose');
    }),
  };

  const parser = new StreamingMessageParser({
    artifactElement: () => '',
    callbacks,
  });

  let message = '';
  let result = '';

  const chunks = Array.isArray(input) ? input : [input];

  for (const chunk of chunks) {
    message += chunk;
    result = parser.parse('message_1', message);
  }

  for (const name in expected.callbacks) {
    const callbackName = name;
    expect(callbacks[callbackName as keyof typeof callbacks]).toHaveBeenCalledTimes(
      expected.callbacks[callbackName as keyof typeof expected.callbacks] ?? 0,
    );
  }

  expect(result).toEqual(expected.output);
}
