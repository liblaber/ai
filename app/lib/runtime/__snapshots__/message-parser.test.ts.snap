// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`StreamingMessageParser valid artifacts with actions should correctly parse chunks and strip out liblab artifacts (0): onActionClose 1`] = `
{
  "action": {
    "content": "npm install",
    "type": "shell",
  },
  "actionId": "0",
  "artifactId": "artifact_1",
  "messageId": "message_1",
  "shouldExecute": true,
}
`;

exports[`StreamingMessageParser valid artifacts with actions should correctly parse chunks and strip out liblab artifacts (0): onActionOpen 1`] = `
{
  "action": {
    "content": "",
    "type": "shell",
  },
  "actionId": "0",
  "artifactId": "artifact_1",
  "messageId": "message_1",
  "shouldExecute": true,
}
`;

exports[`StreamingMessageParser valid artifacts with actions should correctly parse chunks and strip out liblab artifacts (0): onArtifactClose 1`] = `
{
  "id": "artifact_1",
  "messageId": "message_1",
  "title": "Some title",
  "type": undefined,
}
`;

exports[`StreamingMessageParser valid artifacts with actions should correctly parse chunks and strip out liblab artifacts (0): onArtifactOpen 1`] = `
{
  "id": "artifact_1",
  "messageId": "message_1",
  "title": "Some title",
  "type": undefined,
}
`;

exports[`StreamingMessageParser valid artifacts without actions should correctly parse chunks and strip out liblab artifacts (0): onArtifactClose 1`] = `
{
  "id": "artifact_1",
  "messageId": "message_1",
  "title": "Some title",
  "type": undefined,
}
`;

exports[`StreamingMessageParser valid artifacts without actions should correctly parse chunks and strip out liblab artifacts (0): onArtifactOpen 1`] = `
{
  "id": "artifact_1",
  "messageId": "message_1",
  "title": "Some title",
  "type": undefined,
}
`;

exports[`StreamingMessageParser valid artifacts without actions should correctly parse chunks and strip out liblab artifacts (1): onArtifactClose 1`] = `
{
  "id": "artifact_1",
  "messageId": "message_1",
  "title": "Some title",
  "type": undefined,
}
`;

exports[`StreamingMessageParser valid artifacts without actions should correctly parse chunks and strip out liblab artifacts (1): onArtifactOpen 1`] = `
{
  "id": "artifact_1",
  "messageId": "message_1",
  "title": "Some title",
  "type": undefined,
}
`;
